File: binding/src/main/java/xyz/tynn/astring/binding/AStringBinding.java
//  Copyright 2026 Christian Schmitz
//  SPDX-License-Identifier: Apache-2.0

package xyz.tynn.astring.binding;

import static xyz.tynn.astring.AStringFactory.createFromCharSequence;
import static xyz.tynn.astring.AStringKt.invokeWithContext;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.databinding.InverseMethod;

import xyz.tynn.astring.AString;

/**
 * AString converter methods to support two-way data binding
 */
public interface AStringBinding {

    @InverseMethod("wrap")
    static String load(@NonNull Context context, AString value) {
        CharSequence string = invokeWithContext(context, value);
        return string == null ? null : string.toString();
    }

    @NonNull
    static AString wrap(@SuppressWarnings("unused") Context context, String value) {
        return createFromCharSequence(value);
    }
}

File: binding/src/main/kotlin/xyz/tynn/astring/binding/AStringBinding.kt
//  Copyright 2023 Christian Schmitz
//  SPDX-License-Identifier: Apache-2.0

@file:JvmName("AStringBinding")

package xyz.tynn.astring.binding

import android.content.Context
import androidx.databinding.InverseMethod
import xyz.tynn.astring.AString
import java.util.*
import java.util.Objects.requireNonNull

/**
 * AString converter methods to support two-way data binding
 */

@InverseMethod("wrap")
public fun Context.load(
    value: AString?,
): String? = value?.invoke(
    requireNonNull(this, "context"),
)?.toString()

@Suppress("unused")
public fun Context.wrap(
    value: String?,
): AString = AString(value)
